yew/src/sgml_tags.rs

```diff
@@ -0,0 +1,214 @@
+/// in [the WhatWG spec](https://html.spec.whatwg.org/multipage/syntax.html#syntax-attribute-name)
+pub fn is_valid_html_attribute_name(attr: &str) -> bool {
+    for c in attr.chars() {
+        if is_noncharacter(c)
+            || is_control(c)
+            || c == '\u{0020}'
+            || c == '\u{0022}'
+            || c == '\u{0027}'
+            || c == '\u{003E}'
+            || c == '\u{002F}'
+            || c == '\u{003D}'
+        {
+            return false;
+        }
+    }
+    true
+}
+
+/// Returns true when the character provided is a valid PCENChar as defined
+/// in [the WhatWG spec](https://html.spec.whatwg.org/multipage/custom-elements.html#prod-pcenchar)
+fn is_pcen_char(c: char) -> bool {
+    match c {
+        '-' | '.' | '0'..='9' | 'a'..='z' | '_' => true,
+        '\u{B7}' => true,
+        '\u{C0}'..='\u{D6}' => true,
+        '\u{D8}'..='\u{F6}' => true,
+        '\u{F8}'..='\u{37D}' => true,
+        '\u{37F}'..='\u{1FFF}' => true,
+        '\u{200C}'..='\u{200D}' => true,
+        '\u{203F}'..='\u{2040}' => true,
+        '\u{2070}'..='\u{218F}' => true,
+        '\u{2C00}'..='\u{2FEF}' => true,
+        '\u{3001}'..='\u{D7FF}' => true,
+        '\u{F900}'..='\u{FDCF}' => true,
+        '\u{FDF0}'..='\u{FFFD}' => true,
+        '\u{10000}'..='\u{EFFFF}' => true,
+        _ => false,
+    }
+}
+
+/// Returns true when the tag name provided would be a valid "custom element" per
+/// [the WhatWG spec](https://html.spec.whatwg.org/multipage/custom-elements.html#valid-custom-element-name).
+/// Only technically returns correct results if called with a string that is not one of the following:
+///     - annotation-xml
+///     - color-profile
+///     - font-face
+///     - font-face-src
+///     - font-face-uri
+///     - font-face-format
+///     - font-face-name
+///     - missing-glyph
+/// But, given the way it is used in this file, as of this writing, this limitation does not affect the
+/// behavior of the program.
+fn is_valid_html_custom_element_name(tag: &str) -> bool {
+    let mut chars = tag.chars();
+    let first_char = chars.next();
+
+    match first_char {
+        None => false,
+        Some(first_char) => {
+            // must begin with [a-z]
+            if first_char < 'a' || first_char > 'z' {
+                return false;
+            }
+
+            let mut seen_hyphen = false;
+            for c in chars {
+                if c == '-' {
+                    seen_hyphen = true
+                }
+
+                // all characters must be valid PCENChar's
+                if !is_pcen_char(c) {
+                    return false;
+                }
+            }
+
+            // must contain at least one hyphen
+            seen_hyphen
+        }
+    }
+}
+
+/// Returns true when the tag name provided looks like a valid non-custom HTML element or valid SVG element.
+/// There's no official spec here, it's just arbitrary.
+fn is_valid_html_standard_ish_element_name(tag: &str) -> bool {
```

Comment:
Is it just me, or is this function name a bit to long?
I also feel "standard_ish" is a bit confusing.